File: frontend/src/types/operations.ts
export type VesselStatus = 'Arriving' | 'At Anchor' | 'Berthing' | 'Loading' | 'Delayed' | 'Completed';
export type PriorityLevel = 'Standard' | 'Priority' | 'Critical' | 'Urgent';

export interface VesselTimelineEvent {
  stage: string;
  time: string;
  status: 'completed' | 'current' | 'scheduled';
}

export interface Vessel {
  id: string;
  name: string;
  imo: string;
  flag: string;
  lengthMeters: number;
  draughtMeters: number;
  teuCapacity: number;
  cargoVolume: number; // TEU currently loaded/discharging
  origin: string;
  destination: string;
  eta: string;
  etd: string;
  status: VesselStatus;
  priority: PriorityLevel;
  currentBerth: string | null;
  assignedBerth: string;
  predictedWaitHours: number;
  demurrageRisk: 'Low' | 'Medium' | 'High';
  historicalTurnaroundHours: number;
  timelineEvents: VesselTimelineEvent[];
  recommendedAction?: string;
  // Ship Agent Ownership & Voyage Details
  ownerId?: string; // 'demo-agent' or other agency user id
  shippingCompany?: string; // e.g. 'Apex Maritime Lines'
  callSign?: string;
  vesselType?: string; // e.g. 'Container Ship (ULCV)', 'Panamax Feeder'
  voyageNumber?: string;
  previousPort?: string;
  nextPort?: string;
  requestedBerth?: string;
  requestedArrivalTime?: string;
  berthDurationHours?: number;
  requestedCranes?: number;
  cargoType?: string;
  cargoQuantity?: string;
  containersLoaded?: number;
  containersTotal?: number;
  dangerousGoods?: boolean;
  specialNotes?: string;
}

export type BerthStatus = 'Occupied' | 'Available' | 'Congested' | 'Maintenance';

export interface BerthHourlyForecast {
  hour: string;
  utilization: number;
}

export interface Berth {
  id: string;
  name: string;
  lengthMeters: number;
  depthMeters: number;
  maxDraftMeters: number;
  status: BerthStatus;
  currentUtilization: number;
  predictedUtilization: number;
  queueCount: number;
  availableCranes: number;
  maxCranes: number;
  currentVesselId: string | null;
  nextVesselId: string | null;
  assignedCraneIds: string[];
  hourlyForecast: BerthHourlyForecast[];
  riskLevel: 'LOW' | 'MEDIUM' | 'HIGH';
}

export type CraneStatus = 'ACTIVE' | 'IDLE' | 'MAINTENANCE' | 'FAILED';

export interface Crane {
  id: string;
  name: string;
  type: 'STS' | 'Yard' | 'Mobile';
  status: CraneStatus;
  berthId: string;
  movesPerHour: number;
  utilizationPercent: number;
  failureDurationHours?: number;
  impactSeverity?: 'Low' | 'Medium' | 'High' | 'Critical';
  assignedVesselId?: string | null;
  lastMaintenance: string;
  nextMaintenance: string;
}

export interface YardBlock {
  id: string;
  name: string;
  category: 'Dry' | 'Reefer' | 'Hazmat' | 'Empty';
  totalTeu: number;
  occupiedTeu: number;
  utilizationPercent: number;
  dwellTimeDays: number;
  inboundTeu24h: number;
  outboundTeu24h: number;
  congestionRisk: 'Normal' | 'Moderate' | 'High' | 'Severe';
  suggestedRedistribution?: {
    targetBlockId: string;
    amountTeu: number;
    rationale: string;
  };
}

export interface ForecastDriver {
  factor: string;
  percentage: number;
  impact: 'High' | 'Medium' | 'Low';
  detail: string;
}

export interface CongestionForecastData {
  timeHorizon: '6h' | '12h' | '24h' | '48h' | '72h';
  overallCongestionPercent: number;
  riskBottleneckBerth: string;
  bottleneckConfidence: number;
  drivers: ForecastDriver[];
  points: {
    hour: string;
    B01: number;
    B02: number;
    B03: number;
    B04: number;
    B05: number;
    B06: number;
    threshold: number;
  }[];
}

export interface OptimizationPlanDetail {
  vesselId: string;
  vesselName: string;
  berthId: string;
  cranesAssigned: number;
  expectedWaitHours: number;
  berthUtilizationPercent: number;
}

export interface OptimizationResult {
  id: string;
  timestamp: string;
  isApplied: boolean;
  targetVesselId: string;
  currentPlan: OptimizationPlanDetail;
  optimizedPlan: OptimizationPlanDetail;
  waitTimeDeltaHours: number;
  waitTimeReductionPercent: number;
  queueReductionCount: number;
  savingsEstimateUsd: number;
  rationale: string;
}

export interface SimulationScenario {
  id: string;
  type: 'crane_failure' | 'berth_closure' | 'vessel_surge' | 'vessel_delay' | 'yard_capacity_reduction';
  targetEntityId: string;
  targetEntityName: string;
  durationHours: number;
  description: string;
}

export interface SimulationStepAction {
  order: number;
  title: string;
  detail: string;
  targetEntity: string;
}

export interface SimulationResult {
  id: string;
  scenario: SimulationScenario;
  before: {
    queueCount: number;
    avgWaitHours: number;
    berthUtilizationPercent: number;
  };
  after: {
    queueCount: number;
    avgWaitHours: number;
    berthUtilizationPercent: number;
  };
  recoveryPlan: {
    id: string;
    steps: SimulationStepAction[];
    isApplied: boolean;
    expectedRecoveryHours: number;
  };
}

export interface RouteOption {
  portCode: string;
  portName: string;
  country: string;
  distanceNm: number;
  eta: string;
  delayHours: number;
  extraCostUsd: number;
  riskLevel: 'Low' | 'Medium' | 'High' | 'Critical';
  isRecommended: boolean;
  congestionScore: number;
  rationale: string;
  coordinates: { x: number; y: number };
}

export interface ShiftPlanItem {
  id: string;
  dayOffset: number; // 0 = Today, 1 = Tomorrow, 2 = +2 Days, 3 = +3 Days
  shift: '06-14' | '14-22' | '22-06';
  berthId: string;
  vesselId: string | null;
  vesselName: string | null;
  assignedCranes: string[];
  status: 'Scheduled' | 'Conflict' | 'Optimized' | 'Maintenance';
  conflictReason?: string;
}

export interface OperationalAlert {
  id: string;
  timestamp: string;
  severity: 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'INFO';
  title: string;
  description: string;
  relatedEntity: {
    type: 'berth' | 'vessel' | 'crane' | 'yard';
    id: string;
    name: string;
  };
  isResolved: boolean;
  actionRoute: string;
  actionLabel: string;
}

export interface CopilotTableBlock {
  title?: string;
  columns: string[];
  rows: (string | number)[][];
}

export interface CopilotChartBlock {
  chartType: 'bar' | 'line' | 'pie';
  title?: string;
  data: { name: string; value: number; [key: string]: any }[];
  dataKey?: string;
  categoryKey?: string;
}

export interface CopilotMetricItem {
  label: string;
  value: string | number;
  subtext?: string;
  isPositive?: boolean;
}

export interface CopilotStructuredBlock {
  type: 'table' | 'chart' | 'metrics';
  title?: string;
  columns?: string[];
  rows?: (string | number)[][];
  chartType?: 'bar' | 'line' | 'pie';
  data?: { name: string; value: number; [key: string]: any }[];
  metrics?: CopilotMetricItem[];
}

export interface CopilotStructuredResponse {
  message: string;
  blocks?: CopilotStructuredBlock[];
  suggestedActions?: {
    label: string;
    actionRoute?: string;
    prompt?: string;
    requiresConfirmation?: boolean;
  }[];
}

export interface CopilotMessage {
  id: string;
  sender: 'user' | 'gemini';
  text: string;
  timestamp: string;
  structured?: CopilotStructuredResponse;
  toolCalls?: {
    toolName: string;
    parameters?: Record<string, any>;
    resultSummary: string;
  }[];
  suggestedActions?: {
    label: string;
    actionRoute?: string;
    actionType?: string;
    payload?: any;
    prompt?: string;
    requiresConfirmation?: boolean;
  }[];
  isStreaming?: boolean;
}

export type BerthRequestStatus = 'Pending' | 'Under Review' | 'Approved' | 'Changed' | 'Rejected';

export interface BerthRequest {
  id: string;
  vesselId: string;
  vesselName: string;
  imo: string;
  requestedBerth: string;
  requestedArrivalTime: string;
  estimatedDurationHours: number;
  requestedCranes: number;
  cargoType: string;
  status: BerthRequestStatus;
  submittedAt: string;
  reviewedAt?: string;
  assignedBerth?: string;
  notes?: string;
  ownerId: string;
}

export type DocumentType =
  | 'Bill of Lading'
  | 'Cargo Manifest'
  | 'Vessel Declaration'
  | 'Arrival Notice'
  | 'Dangerous Goods Declaration'
  | 'Customs Clearance'
  | 'Crew List'
  | 'Other';

export interface ShippingDocument {
  id: string;
  name: string;
  vesselId: string;
  vesselName: string;
  type: DocumentType;
  status: 'Approved' | 'Under Review' | 'Required' | 'Draft';
  uploadedDate: string;
  fileSize: string;
  fileUrl?: string;
  ownerId: string;
}

File: src/backend/app/models/operations.py
from typing import List, Optional, Literal, Dict, Any
from pydantic import BaseModel, Field

# --- Vessel Models ---
VesselStatus = Literal['Arriving', 'At Anchor', 'Berthing', 'Loading', 'Delayed', 'Completed']
PriorityLevel = Literal['Standard', 'Priority', 'Critical', 'Urgent']

class VesselTimelineEvent(BaseModel):
    stage: str
    time: str
    status: Literal['completed', 'current', 'scheduled']

class Vessel(BaseModel):
    id: str
    name: str
    imo: str
    flag: str
    lengthMeters: float
    draughtMeters: float
    teuCapacity: int
    cargoVolume: int
    origin: str
    destination: str
    eta: str
    etd: str
    status: VesselStatus
    priority: PriorityLevel
    currentBerth: Optional[str] = None
    assignedBerth: str
    predictedWaitHours: float
    demurrageRisk: Literal['Low', 'Medium', 'High']
    historicalTurnaroundHours: float
    timelineEvents: List[VesselTimelineEvent] = []
    recommendedAction: Optional[str] = None

# --- Berth Models ---
BerthStatus = Literal['Occupied', 'Available', 'Congested', 'Maintenance']

class BerthHourlyForecast(BaseModel):
    hour: str
    utilization: int

class Berth(BaseModel):
    id: str
    name: str
    lengthMeters: float
    depthMeters: float
    maxDraftMeters: float
    status: BerthStatus
    currentUtilization: int
    predictedUtilization: int
    queueCount: int
    availableCranes: int
    maxCranes: int
    currentVesselId: Optional[str] = None
    nextVesselId: Optional[str] = None
    assignedCraneIds: List[str] = []
    hourlyForecast: List[BerthHourlyForecast] = []
    riskLevel: Literal['LOW', 'MEDIUM', 'HIGH']

# --- Crane Models ---
CraneStatus = Literal['ACTIVE', 'IDLE', 'MAINTENANCE', 'FAILED']

class Crane(BaseModel):
    id: str
    name: str
    type: Literal['STS', 'Yard', 'Mobile']
    status: CraneStatus
    berthId: str
    movesPerHour: int
    utilizationPercent: int
    failureDurationHours: Optional[int] = None
    impactSeverity: Optional[Literal['Low', 'Medium', 'High', 'Critical']] = None
    assignedVesselId: Optional[str] = None
    lastMaintenance: str
    nextMaintenance: str

# --- Yard Models ---
class YardRedistribution(BaseModel):
    targetBlockId: str
    amountTeu: int
    rationale: str

class YardBlock(BaseModel):
    id: str
    name: str
    category: Literal['Dry', 'Reefer', 'Hazmat', 'Empty']
    totalTeu: int
    occupiedTeu: int
    utilizationPercent: int
    dwellTimeDays: float
    inboundTeu24h: int
    outboundTeu24h: int
    congestionRisk: Literal['Normal', 'Moderate', 'High', 'Severe']
    suggestedRedistribution: Optional[YardRedistribution] = None

# --- Congestion Forecast Models ---
class ForecastDriver(BaseModel):
    factor: str
    percentage: int
    impact: Literal['High', 'Medium', 'Low']
    detail: str

class ForecastPoint(BaseModel):
    hour: str
    B01: int
    B02: int
    B03: int
    B04: int
    B05: int
    B06: int
    threshold: int = 85

class CongestionForecastData(BaseModel):
    timeHorizon: Literal['6h', '12h', '24h', '48h', '72h']
    overallCongestionPercent: int
    riskBottleneckBerth: str
    bottleneckConfidence: int
    drivers: List[ForecastDriver]
    points: List[ForecastPoint]

# --- Optimization Models ---
class OptimizationPlanDetail(BaseModel):
    vesselId: str
    vesselName: str
    berthId: str
    cranesAssigned: int
    expectedWaitHours: float
    berthUtilizationPercent: int

class OptimizationResult(BaseModel):
    id: str
    timestamp: str
    isApplied: bool
    targetVesselId: str
    currentPlan: OptimizationPlanDetail
    optimizedPlan: OptimizationPlanDetail
    waitTimeDeltaHours: float
    waitTimeReductionPercent: float
    queueReductionCount: int
    savingsEstimateUsd: float
    rationale: str

class OptimizationRequest(BaseModel):
    vesselId: Optional[str] = None
    prioritizeDemurrage: bool = True

# --- Disruption Simulation Models ---
class SimulationScenario(BaseModel):
    id: str
    type: Literal['crane_failure', 'berth_closure', 'vessel_surge', 'vessel_delay', 'yard_capacity_reduction']
    targetEntityId: str
    targetEntityName: str
    durationHours: int
    description: str

class SimulationStepAction(BaseModel):
    order: int
    title: str
    detail: str
    targetEntity: str

class SimulationMetrics(BaseModel):
    queueCount: int
    avgWaitHours: float
    berthUtilizationPercent: int

class SimulationRecoveryPlan(BaseModel):
    id: str
    steps: List[SimulationStepAction]
    isApplied: bool
    expectedRecoveryHours: float

class SimulationResult(BaseModel):
    id: str
    scenario: SimulationScenario
    before: SimulationMetrics
    after: SimulationMetrics
    recoveryPlan: SimulationRecoveryPlan

# --- Route Models ---
class RouteCoordinates(BaseModel):
    x: float
    y: float

class RouteOption(BaseModel):
    portCode: str
    portName: str
    country: str
    distanceNm: int
    eta: str
    delayHours: float
    extraCostUsd: int
    riskLevel: Literal['Low', 'Medium', 'High', 'Critical']
    isRecommended: bool
    congestionScore: int
    rationale: str
    coordinates: RouteCoordinates

# --- Shift Planner Models ---
class ShiftPlanItem(BaseModel):
    id: str
    dayOffset: int
    shift: Literal['06-14', '14-22', '22-06']
    berthId: str
    vesselId: Optional[str] = None
    vesselName: Optional[str] = None
    assignedCranes: List[str] = []
    status: Literal['Scheduled', 'Conflict', 'Optimized', 'Maintenance']
    conflictReason: Optional[str] = None

# --- Alert Models ---
class AlertEntity(BaseModel):
    type: Literal['berth', 'vessel', 'crane', 'yard']
    id: str
    name: str

class OperationalAlert(BaseModel):
    id: str
    timestamp: str
    severity: Literal['CRITICAL', 'HIGH', 'MEDIUM', 'INFO']
    title: str
    description: str
    relatedEntity: AlertEntity
    isResolved: bool
    actionRoute: str
    actionLabel: str

# --- Port Status Summary ---
class PortStatusSummary(BaseModel):
    vesselsInPort: int
    vesselsArriving: int
    activeBerthUtilization: int
    craneAvailability: int
    yardUtilization: int
    currentQueue: int
    predictedCongestion: Literal['NORMAL', 'HIGH', 'CRITICAL']
    highRiskBerth: str
    highRiskBerthUtilization: int
    highRiskBerthPredicted: int
